### PR TITLE
8302068: Serial: Refactor oop closures used in Young GC

### PR DESCRIPTION
**Pull request by albertnetymk**
The diff can be a bit hard to follow; it's best to view new code on its own.

Test: tier1-5

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302068](https://bugs.openjdk.org/browse/JDK-8302068): Serial: Refactor oop closures used in Young GC

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [3eca051f](https://git.openjdk.org/jdk/pull/12471/files/3eca051ff1d5e578a35521bef739ffee999d1970)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**) ⚠️ Review applies to [c51fffa9](https://git.openjdk.org/jdk/pull/12471/files/c51fffa9e42824fc74ca5e83bb29baf77520d1a6)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12471/head:pull/12471` \
`$ git checkout pull/12471`

Update a local copy of the PR: \
`$ git checkout pull/12471` \
`$ git pull https://git.openjdk.org/jdk pull/12471/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12471`

View PR using the GUI difftool: \
`$ git pr show -t 12471`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12471.diff">https://git.openjdk.org/jdk/pull/12471.diff</a>

</details>

### CODE REVIEW

**Review by tschatzl — APPROVED** [review]
Lgtm.

**Review by walulyai — APPROVED** [review]
Lgtm!

Minor:
Few uses of NULL instead of nullptr.
